File: Algorithms/Merging-Intervals/merge_intervals.cpp
#include<bits/stdc++.h>
using namespace std;
 
// An interval has start time and end time
struct Interval
{
    int start, end;
};
 
// To compare two intervals accoridng to their start time
bool myComp(Interval interval1, Interval interval2)
{
    return (interval1.start < interval2.start);
}
 
// This function takes a set of intervals, merges
// overlapping intervals and prints the result
void mergeIntervals(Interval arr[], int n)
{
    // Test if the given set has at least one interval
    if (n <= 0)
        return;
 
    // Create an empty stack of intervals
    stack<Interval> s;
 
    // sorting the intervals in increasing order of start time
    sort(arr, arr+n, myComp);
 
    // push the first interval to stack
    s.push(arr[0]);
 
    // Start from the next interval and merge if necessary
    for (int i = 1 ; i < n; i++)
    {
        // get interval from stack top
        Interval top = s.top();
 
        // if current interval is not overlapping with stack top,
        // push it to the stack
        if (top.end < arr[i].start)
            s.push(arr[i]);
 
        // Otherwise update the ending time of top if ending of current
        // interval is more
        else if (top.end < arr[i].end)
        {
            top.end = arr[i].end;
            s.pop();
            s.push(top);
        }
    }
 
    // Print contents of stack
    cout << "\n After merging Intervals are: ";
    while (!s.empty())
    {
        Interval t = s.top();
        cout << "[" << t.start << "," << t.end << "] ";
        s.pop();
    }
    return;
}
int main()
{
  Interval arr[]= {{1,6},{3,9},{11,13},{2,5}};  //creating an array
    int n = sizeof(arr)/sizeof(arr[0]); //size of the array
    mergeIntervals(arr,n);
    return 0;
}

File: Algorithms/Merging-Intervals/merge_intervals.js
// list of intervals as an array of objects
const arr = [
    { start: 1, end: 3 },
    { start: 2, end: 4 },
    { start: 5, end: 8 },
    { start: 6, end: 7 }
];


function mergeIntervals(arr) {
    // sort the array of intervals in ascending order based on the start value of each interval
    arr.sort(function(a, b) {
        if (a.start < b.start) return -1;
        if (a.start > b.start) return 1;
        return 0;
    });

    // result array which will work as stack
    let res = [];

    // push first interval into result array
    res.push(arr[0]);

    for (let i = 1; i < arr.length; i++) {
        // get the top of the result array
        // we will call it top interval
        const top = res[res.length - 1]

        // top interval's end is less than the current interval's start that means disjoint intervals
        // we can push the current interval to our result
        if (top.end < arr[i].start) {
            res.push(arr[i]);
        }
        // top interval's end is between the current interval's start and end
        // we can merge intervals with the ending at current interval's end
        else if (top.end < arr[i].end) {
            top.end = arr[i].end;
            // delete the top interval
            res.pop();
            // insert new modified interval
            res.push(top);
        }
        // Otherwise top interval's end is after the current interval's end
        // i.e after merge we will get top interval, hence no change required
    }

    // print the array of merged intervals
    console.log(res);
}

mergeIntervals(arr);